ssl/quic/quic_txp.c

```diff
@@ -0,0 +1,2027 @@
+    h->bytes_appended       = 0;
+    h->scratch_bytes        = 0;
+    h->reserve_allowed      = 0;
+    h->done_implicit        = 0;
+    h->txn.buf_mem.data     = NULL;
+    h->txn.buf_mem.length   = 0;
+    h->txn.buf_mem.max      = 0;
+    h->txn.active           = 0;
+
+    if (max_ppl > h->txp->scratch_len) {
+        unsigned char *scratch;
+
+        scratch = OPENSSL_realloc(h->txp->scratch, max_ppl);
+        if (scratch == NULL)
+            return 0;
+
+        h->txp->scratch     = scratch;
+        h->txp->scratch_len = max_ppl;
+    }
+
+    return 1;
+}
+
+static void tx_helper_cleanup(struct tx_helper *h)
+{
+    if (h->txn.active)
+        tx_helper_rollback(h);
+
+    h->txp = NULL;
+}
+
+static void tx_helper_unrestrict(struct tx_helper *h)
+{
+    h->reserve_allowed = 1;
+}
+
+/*
+ * Append an extent of memory to the iovec list. The memory must remain
+ * allocated until we finish generating the packet and call the QTX.
+ *
+ * In general, the buffers passed to this function will be from one of two
+ * ranges:
+ *
+ *   - Application data contained in stream buffers managed elsewhere
+ *     in the QUIC stack; or
+ *
+ *   - Control frame data appended into txp->scratch using tx_helper_begin and
+ *     tx_helper_commit.
+ *
+ */
+static int tx_helper_append_iovec(struct tx_helper *h,
+                                  const unsigned char *buf,
+                                  size_t buf_len)
+{
+    if (buf_len == 0)
+        return 1;
+
+    if (!ossl_assert(!h->done_implicit))
+        return 0;
+
+    if (!txp_ensure_iovec(h->txp, h->num_iovec + 1))
+        return 0;
+
+    h->txp->iovec[h->num_iovec].buf     = buf;
+    h->txp->iovec[h->num_iovec].buf_len = buf_len;
+
+    ++h->num_iovec;
+    h->bytes_appended += buf_len;
+    return 1;
+}
+
+/*
+ * How many more bytes of space do we have left in our plaintext packet payload?
+ */
+static size_t tx_helper_get_space_left(struct tx_helper *h)
+{
+    return h->max_ppl
+        - (h->reserve_allowed ? 0 : h->reserve) - h->bytes_appended;
+}
+
+/*
+ * Begin a control frame serialization transaction. This allows the
+ * serialization of the control frame to be backed out if it turns out it won't
+ * fit. Write the control frame to the returned WPACKET. Ensure you always
+ * call tx_helper_rollback or tx_helper_commit (or tx_helper_cleanup). Returns
+ * NULL on failure.
+ */
+static WPACKET *tx_helper_begin(struct tx_helper *h)
+{
+    size_t space_left;
+
+    if (!ossl_assert(!h->txn.active))
+        return NULL;
+
+    if (!ossl_assert(!h->done_implicit))
+        return NULL;
+
+    h->txn.buf_mem.data     = (char *)h->txp->scratch + h->scratch_bytes;
+    h->txn.buf_mem.length   = h->txp->scratch_len - h->scratch_bytes;
+    h->txn.buf_mem.max      = h->txn.buf_mem.length;
+
+    space_left = tx_helper_get_space_left(h);
+    if (!ossl_assert(space_left <= h->txn.buf_mem.max))
+        return NULL;
+
+    if (!WPACKET_init(&h->txn.wpkt, &h->txn.buf_mem))
```

Comment:
Would `WPACKET_init_static_len` be more appropriate here? If passed a `BUF_MEM` the WPACKET code may call `BUF_MEM_grow` if it ever runs out of space. This will attempt to reallocate the underlying buffer....but that doesn't seem appropriate since we are using the scratch buffer.

Reply:
Good catch.